scrapy/downloadermiddlewares/httpauth.py

```diff
@@ -26,6 +26,14 @@ def spider_opened(self, spider):
             self.auth = basic_auth_header(usr, pwd)
 
     def process_request(self, request, spider):
-        auth = getattr(self, 'auth', None)
-        if auth and b'Authorization' not in request.headers:
+        if b'Authorization' in request.headers:
+            return
+
+        usr = request.meta.get('http_user', '')
+        pwd = request.meta.get('http_pass', '')
+        if usr or pwd:
+            auth = basic_auth_header(usr, pwd)
+        else:
+            auth = getattr(self, 'auth', None)
+        if auth:
```

Comment:
I moved that part to the beginning of the method, so that if there is already such a header we save the CPU needed for the rest of the logic (e.g. `basic_auth_header`).

Reply:
Oh nevermind then, didn't see that :facepalm: